### ISSUE TITLE
Header is missing in the GTrack file

### ISSUE THREAD

**Issue by QianzhaoJ** [Chrom3D/Chrom3D#10]
Hi:
I follow the pipeline to create the input Model Setup File (GTrack) from the Hi-C data , and  I have got the file `sample_inter_intra_chr.diploid.gtrack` . But when I run Chrom3D to generate 3D genome models , I got this error : 
> terminate called after throwing an instance of 'std::runtime_error'
    what():  Header is missing in the GTrack file
    Aborted (core dumped)

And the header of my GTrack file looks like:

> ##gtrack version: 1.0
##track type: linked segments
###seqid        start   end     id      radius  periphery       edges
chr1_A  0       3950000 chr1_A:0-3950000        1       .
chr1_B  0       3950000 chr1_B:0-3950000        1       .
chr1_A  3950000 6000000 chr1_A:3950000-6000000  1       .
chr1_B  3950000 6000000 chr1_B:3950000-6000000  1       .
chr1_A  6000000 14300000        chr1_A:6000000-14300000 1       .
chr1_B  6000000 14300000        chr1_B:6000000-14300000 1       .

So , could you give me some suggestions ?

Thank you!
Best wishes
Qianzhao

**Comment by QianzhaoJ** [comment]
Hi :
Maybe this is because I didn't add LAD information to the GTrack file ,so the header `periphery` should not be in GTrack file , then I remove it and it works .
Is that right?

Best wishes
Qianzhao

**Comment by sespesogil** [comment]
I have the same problem. I didnt add LADs, but my diploid.gtracks looks different: 

##gtrack version: 1.0
##track type: linked segments
###seqid        start   end     id      radius  edges
chr1    0       3000000 chr1:0-3000000  1       .
chr1    3000000 4360000 chr1:3000000-4360000    1       chr1:84700000-85830000;chr1:88200000-88350000

**Comment by sespesogil** [comment]
If I remove the edges is totally doing something, but it finishes crashing with boost complaining

Chrom3D: /boost/1.60.0/include/boost/random/uniform_int_distribution.hpp:337: boost::random::uniform_int_distribution<IntType>::uniform_int_distribution(IntType, IntType) [with IntType = int]: Assertion `min_arg <= max_arg' failed.
Aborted

**Comment by sespesogil** [comment]
Maybe the ids? shouldn't be that distinct? They are all 1

**Comment by tharvesh** [comment]
> Hi :
> Maybe this is because I didn't add LAD information to the GTrack file ,so the header `periphery` should not be in GTrack file , then I remove it and it works .
> Is that right?
> 
> Best wishes
> Qianzhao

Hi Qianzhao,
That's right. If you don't have LAD data then please use the script "add_inter_chrom_beads_wo_lads.sh" while adding inter-chromosomal interactions.

Cheers
Tharvesh

**Comment by tharvesh** [comment]
Hi sespesogil,

IDs are distinct in your file (fourth field). From what you pasted as the "diploid Gtrack", it is not a diploid Gtrack. Could you please paste the error message when you executed Chrom3D without removing "edges" field.

Thanks
Tharvesh

**Comment by sespesogil** [comment]
Yes, sorry. I was kind of excited wanting to see the results and I pasted another gtrack by error. I had the exact same error as Qianzhao, but I just run the add_inter_chrom_beads_wo_lads.sh . My diploid gtrack looks like that: 

##gtrack version: 1.0
##track type: linked segments
###seqid        start   end     id      radius  edges
chr1_A  0       3000000 chr1_A:0-3000000        1       .
chr1_B  0       3000000 chr1_B:0-3000000        1       .
chr1_A  3000000 4360000 chr1_A:3000000-4360000  1       chr1_A:84700000-85830000;chr1_A:88200000-88350000;chr5_A:97110000-98190000;chr5_A:96350000-96710000;chr5_A:103840000-104700000;chr4_A:114300000-114630000;chr5_A:0-4200000

Chrom3D seems to work , no header error, but this error:

Chrom3D: /boost/1.60.0/include/boost/random/uniform_int_distribution.hpp:337: boost::random::uniform_int_distribution<IntType>::uniform_int_distribution(IntType, IntType) [with IntType = int]: Assertion `min_arg <= max_arg' failed.
Aborted


What do you think should be the problem?  Thanks a lot Tharvesh!

**Comment by QianzhaoJ** [comment]
Hi Tharvesh:

Thank for your help! But I have some other problems now.
I have got 3D genome structural model using Chrom3D, and used Chimera to visualize the 3D genome model just like describing in Nat protocal . But I find that many beads are not placed within 5 µm nucleus, and Some chromosomes are far away from other chromosomes . Is that normal?
I want to study interchromosomal pairwise distance using Chrom3D ,can you give me some suggestions?
Thank you again, and I look forward to hearing from you.

Best wishes
Qianzhao

发件人: Tharvesh Moideen Liyakat Ali<mailto:notifications@github.com>
发送时间: 2018年10月3日 16:51
收件人: Chrom3D/Chrom3D<mailto:Chrom3D@noreply.github.com>
抄送: QianzhaoJ<mailto:jiqianzhaonotold@outlook.com>; Author<mailto:author@noreply.github.com>
主题: Re: [Chrom3D/Chrom3D] Header is missing in the GTrack file (#10)


Hi :
Maybe this is because I didn't add LAD information to the GTrack file ,so the header periphery should not be in GTrack file , then I remove it and it works .
Is that right?

Best wishes
Qianzhao

Hi Qianzhao,
That's right. If you don't have LAD data then please use the script "add_inter_chrom_beads_wo_lads.sh" while adding inter-chromosomal interactions.

Cheers
Tharvesh

**Comment by sespesogil** [comment]
I have been testing your testfles and everything seem to work fine. When I compare both gtracks I can see that the sorting is different. Your "toy-global-example.gtrack"  jumps from chr1 to chr2, meanwhile my file jumps to chr10 instead. Other than that... I dont see anyother difference. 

Help appreaciated. Thanks a lot!!! 

S.

**Comment by sespesogil** [comment]
This is the full error message:

chromosomes: 42
beads: 13006
interactions: 17488
interactions with given weight: 0
non-bounded interactions with given distance: 0
upper-bounded interactions with given distance: 0
lower-bounded interactions with given distance: 0
periphery beads: 0
periphery beads (with weights): 0
center beads: 0
center beads (with weights): 0
0 1.01246e+08
Chrom3D: /boost/1.60.0/include/boost/random/uniform_int_distribution.hpp:337: boost::random::uniform_int_distribution<IntType>::uniform_int_distribution(IntType, IntType) [with IntType = int]: Assertion `min_arg <= max_arg' failed.
Aborted

**Comment by jonnings** [comment]
sespesogil: Could it be that you have included patch chromosomes in your GTrack file (chrM, or _random, etc.)? This error message suggests that you are including some very small chromosomes consisting of too few beads.

**Comment by sespesogil** [comment]
Yeah... I have checked for that. But no...I don't unfortunately.  I am gonna try to run again Arrowheads and any other TAD caller... Anyway I can plot single chromosomes. I am a bit puzzled being honest. Let's see if I find the problem.

Thanks!

**Comment by sespesogil** [comment]
I found the error. You were right, the problem was chrY, it had only a single bead. If I remove it, it is totally fine I can plot the rest of the chromosomes.  I have realized that I run Arrowhead in with predetermined parameters and that is 100kb resolution.  Now rerunning it with 50kb. I guess that is the origin of the problem. Correct me if I am wrong! 

Thanks a lot for the hint!

**Comment by sespesogil** [comment]
Nothing related concerning the TAD calling. For whatever reason chrY in mouse gives problems.

**Comment by zjiang-lji** [comment]
Dear jonnings,

Similar to sespesogil's Gtrack file, mine jumps from chr1 to chr10. I have no patch chromosomes in my GTrack. I have id and edges columns. I'm getting a similar error like the following. I was wondering if you can help me with this.
```
# Rescaled bead sizes from 27.184 to 0.15
# chromosomes: 46
# beads: 3398
# interactions: 6898
# interactions with given weight: 0
# non-bounded interactions with given distance: 0
# upper-bounded interactions with given distance: 0
# lower-bounded interactions with given distance: 0
# periphery beads: 0
# periphery beads (with weights): 0
# center beads: 0
# center beads (with weights): 0
Chrom3D: /share/apps/libs/boost/boost_1_63_0/include/boost/random/uniform_int_distribution.hpp:337: boost::random::uniform_int_distribution<IntType>::uniform_int_distribution(IntType, IntType) [with IntType = int]: Assertion `min_arg <= max_arg' failed.
Aborted (core dumped)
```
My header looks like this:
```
##gtrack version: 1.0
##track type: linked segments
###seqid        start   end     id      radius  edges
chr1_A  0       6650000 chr1_A:0-6650000        1       chr16_A:0-5200000;chr17_A:81100000-83257441;chr5_A:174900000-178450000;chr7_A:71700000-77900000;chr8_A:141300000-145138636;chr2_A:239000000-242193529;chr9_A:132950000-138394717
```
Thank you!

Zichen "Cardiff" Jiang

**Comment by sespesogil** [comment]
Dear Zichen "Cardiff" Jiang,

As Jonas commented this error occurs when you have very few beads in your chromosomes or the size is not right. Check that in your file, you will find right away. If that is the case, try to use different parameters in your TAD caller, that usually solves the issue. 

Hope this helps.

S.

**Comment by zjiang-lji** [comment]
Dear sespesogil,

Thank you for your reply! I appreciate your help. I have some follow-up questions that I hope you could help me with.

I don't think having very few beads is the problem. One Gtrack file that I have access to ran successfully and it has the following beads and interaction counts that are very similar to my Gtrack file's counts. 
```
A Gtrack file that successfully ran:
# Rescaled bead sizes from 125.926 to 0.15
# chromosomes: 40
# beads: 3400
# interactions: 7430
```
```
My Gtrack file that reported the `min_arg <= max_arg' error:
# Rescaled bead sizes from 27.184 to 0.15
# chromosomes: 46
# beads: 3398
# interactions: 6898
```
What do you mean by the size is not right? How do I look for that? Also, I'm using Arrowhead as my TAD caller. What parameters do I change and how do I change them? 50,000 resolution seems to remain the best parameter for my data.
```
arrowhead [-c chromosome(s)] [-m matrix size] [-r resolution] [--threads num_threads]
		[-k normalization (NONE/VC/VC_SQRT/KR)] <HiC file> 
		<output_file>
```

Thank you again! I look forward to your reply.

Zichen "Cardiff" Jiang

**Comment by sespesogil** [comment]
Hi Zichen "Cardiff" Jiang,

I would just aggregate the data in a way it could tell me the number of beads per chromosome. If you have a chromosome with very few beads you will get that error. 

Concerning the size (in bp) of the beads. Well, it is not mutually exclusive with the number of beads per chromosome. The reason is simple, if you have a bead which size is unrealistically too big, then it is a constrain to have more beads in that chromosome. This tends to happen particularly in smaller chromosomes.

You can get directly this information from your Gtrack. 

Hope this helps.

S.

**Comment by zjiang-lji** [comment]
Dear sespesogil,

I'll take a look at that. Thank you so much!

Zichen "Cardiff" Jiang

**Comment by zjiang-lji** [comment]
Dear sespesogil,

I examined my GTrack file and found that chr20 only has two lines (two beads) that span across the entire chr20. This has to be the cause of the min_arg <= max_arg error. You mentioned changing the parameters of TAD callers to fix this problem. Could you elaborate on that? Thank you.

Best regards,
Zichen "Cardiff" Jiang

**Comment by sespesogil** [comment]
That´s awesome. For sanity check you can remove that chromosome and run Chrom3D to see if goes smoothly.  Obviously removing an autosomal chromosome it is not something you might want. Then, you need to play with Arrowhead parameters (or other TAD caller and adapt the format for Chrom3D) . Changing the resolution might help, or the different normalization strategies Arrowhead allows. Try to benchmark that. Probably iterate multiple parameters and then check the results. 

After running Arrowhead, don´t go through Chrom3D directly. It might be time consuming. Try to aggregate the data as you did for your Gtrack with Arrowhead results. Few TADs per chromosome or very big sizes (in bp) will be the no go for Chrom3D.  

Hope this helps!

S.

**Comment by zjiang-lji** [comment]
Dear sespesogil,

Thank you for those tips! I have three follow-up questions.
1) By removing chr20 from chrom.sizes.sorted file and re-running the entire Chrom3D procedure, the purpose is to check if there're other things wrong, correct?
2) Calling Arrowhead with resolution 50Kb produces the largest output file. I should be aiming for the largest possible TAD BED file, correct?
3) My aggregation method for my GTrack is not efficient. Would you mind sharing how you aggregated your GTrack or your Arrowhead result?

Thank you!

Best regards,
Zichen "Cardiff" Jiang

**Comment by sespesogil** [comment]
Answers bellow:

1. Yup. You can remove it right away from the Gtrack. 
2. Not always.  Aggregate the data in a away you can select some sets to run with Chrom3D. There is (I think) a limit in the number of beads Chrom3D can use. 
3.  In R is fairly simple. You just want to count the number of rows that belong to a given chromosome.  You can use for example table , see : https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/table

**Comment by zjiang-lji** [comment]
Dear sespesogil,

1. My first instinct was to remove the two chr20 beads from my GTrack and re-run Chrom3D with this new GTrack. But numerous other beads' edges column refer to the two chr20 beads that I removed. How should I circumvent this problem?
2. If I understood this correctly, do you mean that I should use the Arrowhead BED file with the most number of rows per chromosome?
3. I'll try that. Thank you!

Do you think removing chr20 from chrom.sizes.sorted file and re-running the entire Chrom3D procedure would be another viable way of solving this assertion error?

Best regards,
Zichen "Cardiff" Jiang

**Comment by sespesogil** [comment]
Hi again, 

1. Try to run it by removing chr20. If I remember well it will be ok and the assertion error would not be produced again. Unless you have other problem. 
2. No exactly. Try to select some of the runs based in a reasonable amount of TADs per chromosome. Some times it will work sometimes don't, that is why you need to benchmark it a bit.

**Comment by zjiang-lji** [comment]
Dear sespesogil,

I found that my assertion error is caused by missing the chromosome number "20" in my Arrowhead parameter. I wouldn't have scrutinized it without your help, so thank you so much!

Do you happen to have a script that can convert .cmm file format to .pdb format?

Best regards,
Zichen "Cardiff" Jiang

**Comment by sespesogil** [comment]
That´s awesome!. I am glad you solved it. 

Unfortunately I don´t have any converter  to .pdb format.  :(

All the best, 

S.

**Comment by jonnings** [comment]
@zjiang-lji : If you open the .cmm file in [ChimeraX](https://www.rbvi.ucsf.edu/chimerax/download.html), you can save as .pdb

**Comment by zjiang-lji** [comment]
Dear sespesogil and jonnings,

Perfect. Thank you all!

Best regards,
Zichen "Cardiff" Jiang

**Comment by 532at** [comment]
Hello All, 
Could you provide code for running add_inter_chrom_beads_wo_lads.sh script when no LAD information is present. I have troubles running it. I posted on GitHub, but no solutions until now. https://github.com/Chrom3D/preprocess_scripts/issues/2

Thanks, Adrian

**Comment by Simple53** [comment]
> > Hi :
> > Maybe this is because I didn't add LAD information to the GTrack file ,so the header `periphery` should not be in GTrack file , then I remove it and it works .
> > Is that right?
> > Best wishes
> > Qianzhao
> 
> Hi Qianzhao, That's right. If you don't have LAD data then please use the script "add_inter_chrom_beads_wo_lads.sh" while adding inter-chromosomal interactions.
> 
> Cheers Tharvesh

Hi all，
if I don't have the LAD data, could I generate the 3d structure of the whole genome using this script? How could I run it?
